src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java

```diff
@@ -0,0 +1,636 @@
+            }
+        }
+    }
+
+    private static class MacOSXWatchKey extends AbstractWatchKey {
+        private static final Unsafe unsafe = Unsafe.getUnsafe();
+
+        private static final long kFSEventStreamEventFlagRootChanged     = 0x00000020;
+
+        private static final Path relativeRootPath = theFS.getPath("");
+
+        // Full path to this key's watch root directory.
+        private final Path   realRootPath;
+        private final int    realRootPathLength;
+        private final Object rootPathKey;
+
+        // Kinds of events to be reported.
+        private EnumSet<FSEventKind> eventsToWatch;
+
+        // Native FSEventStreamRef as returned by FSEventStreamCreate().
+        private long         eventStreamRef;
+        private final Object eventStreamRefLock = new Object();
+
+        private DirectorySnapshot directorySnapshot;
+
+        MacOSXWatchKey(final MacOSXWatchService watchService,
+                       final UnixPath dir,
+                       final Object rootPathKey) throws IOException {
+            super(dir, watchService);
+            this.realRootPath       = dir.toRealPath().normalize();
+            this.realRootPathLength = realRootPath.toString().length() + 1;
+            this.rootPathKey        = rootPathKey;
+        }
+
+        synchronized void enable(final CFRunLoopThread runLoopThread,
+                                 final EnumSet<FSEventKind> eventsToWatch,
+                                 final EnumSet<WatchModifier> modifierSet) throws IOException {
+            assert(!isValid());
+
+            this.eventsToWatch = eventsToWatch;
+
+            directorySnapshot = DirectorySnapshot.create(getRealRootPath());
+
+            synchronized (eventStreamRefLock) {
+                final int kFSEventStreamCreateFlagWatchRoot  = 0x00000004;
+                eventStreamRef = MacOSXWatchService.eventStreamCreate(
+                        realRootPath.toString(),
+                        WatchModifier.sensitivityOf(modifierSet),
+                        kFSEventStreamCreateFlagWatchRoot);
+
+                if (eventStreamRef == 0)
+                    throw new IOException("Unable to create FSEventStream");
+
+                MacOSXWatchService.eventStreamSchedule(eventStreamRef,
+                        runLoopThread.getRunLoopRef());
+            }
+        }
+
+        synchronized void disable() {
+            invalidate();
+            directorySnapshot = null;
+        }
+
+        synchronized void handleEvents(final String[] paths, long eventFlagsPtr) {
+            if (paths == null) {
+                reportOverflow(null);
+                return;
+            }
+
+            if (updateNeeded(paths, eventFlagsPtr)) {
+                directorySnapshot.update(MacOSXWatchKey.this);
+            }
+        }
+
+        private Path toRelativePath(final String absPath) {
+            return   (absPath.length() > realRootPathLength)
```

Comment:
Nit: Are there several spaces after `return`?

Reply:
Yes, to align the beginning of the returned expression with its continuations on the subsequent two lines.

Reply:
OK